bundles/org.openhab.voice.googlestt/src/main/java/org/openhab/voice/googlestt/internal/GoogleSTTService.java

```diff
@@ -0,0 +1,381 @@
+    public STTServiceHandle recognize(STTListener sttListener, AudioStream audioStream, Locale locale,
+            Set<String> set) {
+        AtomicBoolean keepStreaming = new AtomicBoolean(true);
+        Future scheduledTask = backgroundRecognize(sttListener, audioStream, keepStreaming, locale, set);
+        return new STTServiceHandle() {
+            @Override
+            public void abort() {
+                keepStreaming.set(false);
+                try {
+                    Thread.sleep(100);
+                } catch (InterruptedException e) {
+                }
+                scheduledTask.cancel(true);
+            }
+        };
+    }
+
+    private void updateConfig() {
+        String clientId = this.config.clientId;
+        String clientSecret = this.config.clientSecret;
+        if (!clientId.isBlank() && !clientSecret.isBlank()) {
+            var oAuthService = oAuthFactory.createOAuthClientService(SERVICE_PID, GCP_TOKEN_URI, GCP_AUTH_URI, clientId,
+                    clientSecret, GCP_SCOPE, false);
+            this.oAuthService = oAuthService;
+            if (!this.config.oauthCode.isEmpty()) {
+                getAccessToken(oAuthService, this.config.oauthCode);
+                deleteAuthCode();
+            }
+        } else {
+            logger.warn("Missing authentication configuration to access Google Cloud STT API.");
+        }
+    }
+
+    private void getAccessToken(OAuthClientService oAuthService, String oauthCode) {
+        logger.debug("Trying to get access and refresh tokens.");
+        try {
+            oAuthService.getAccessTokenResponseByAuthorizationCode(oauthCode, GCP_REDIRECT_URI);
+        } catch (OAuthException | OAuthResponseException e) {
+            if (logger.isDebugEnabled()) {
+                logger.debug("Error fetching access token: {}", e.getMessage(), e);
+            } else {
+                logger.warn("Error fetching access token. Invalid oauth code? Please generate a new one.");
+            }
+        } catch (IOException e) {
+            logger.warn("An unexpected IOException occurred when fetching access token: {}", e.getMessage());
+        }
+    }
+
+    private void deleteAuthCode() {
+        try {
+            org.osgi.service.cm.Configuration serviceConfig = configAdmin.getConfiguration(SERVICE_PID);
+            Dictionary<String, Object> configProperties = serviceConfig.getProperties();
+            if (configProperties != null) {
+                configProperties.put("oauthCode", "");
+                serviceConfig.update(configProperties);
+            }
+        } catch (IOException e) {
+            logger.warn("Failed to delete current oauth code, please delete it manually.");
+        }
+    }
+
+    private Future<?> backgroundRecognize(STTListener sttListener, AudioStream audioStream, AtomicBoolean keepStreaming,
+            Locale locale, Set<String> set) {
+        Credentials credentials = getCredentials();
+        return executor.submit(() -> {
```

Comment:
Is two tasks needed ?
Why not doing like in your procupineks service ?

Reply:
I do not understand this comment. I'm just using one task. Did you mean why I call executor submit inside the function, instead of wrap the method call? I can change it to be a void and call the executor on the recognize method if so.

Reply:
Oups, this is just a wrong comment, a bad reading of the code.
This looks ok as you did.